Support loading Data Values files in the same ways "ordinary" files are loaded

pivotaljohn opened the pull request:
Reusing the strategy by which a filepath from `--file` is used to resolve to the contents of the file(s) for loading Data Values.

All types of inputs are now supported (and handled identically) across `--file`, `--data-value-file`, and `--data-values-file`: 🎉 
- ordinary file
- filesystem directories
- named pipes
- HTTP/HTTPS URLs
- standard input

jtigger committed:
Add E2E for data-values-file reading a dir

jtigger committed:
Accept directory paths in ReadFilesFunc

... so that data values can be loaded via all the strategies that files
are loaded from `--file`

Also:
- update the starting example to illustrate that all values are plain
  merged.
- Share symlink options between files and DV options structs so that
  when the configuration is set, it's available to both
  RegularilesSourceOpts code and DataValuesFlag code.
- Update help text with more broadly used file input formats.
- Adjust some other help text to be more helpful.

cppforlife commented on pkg/cmd/template/data_values_flags.go:
```diff
@@ -149,34 +152,43 @@ func (s *DataValuesFlags) file(fullPath string, strict bool) ([]*datavalues.Enve
 	var result []*datavalues.Envelope
+	for _, dvFile := range dvFiles {
+		// skip non-YAML files that are contained within the specified directory
+		if strings.Contains(dvFile.RelativePath(), "/") && !(dvFile.Type() == files.TypeYAML) {
```
> strings.Contains(dvFile.RelativePath(), "/")

whats this bit about?

pivotaljohn replied:
If the user supplies a non-YAML file directly as the parameter for the flag, we don't want to ignore it. It's only _implied_ files (i.e. contained within a subdirectory) that we want to ignore.

pivotaljohn replied:
... although, I had intended to check if `RelativePath()` was returning a normalized path or if we needed to grab the OS-specific path separator token, here.

I'll do that.

pivotaljohn replied:
_(discussed in [Slack :thread: in #carvel](https://kubernetes.slack.com/archives/CH8KCCKA5/p1653056007351229?thread_ts=1653053084.082349&cid=CH8KCCKA5))_

pivotaljohn replied:
Added `File.IsImplied()` and make check OS-aware.

cppforlife commented on pkg/cmd/template/data_values_flags.go:
```diff
@@ -40,8 +39,10 @@ type DataValuesFlags struct {
+	EnvironFunc   func() []string
+	ReadFilesFunc func(paths string) ([]*files.File, error)
+
+	files.SymlinkAllowOpts
```
since users control the "values pipe" i dont think we should be restricting usage of symlinks for data value ingestion.

pivotaljohn replied:
I had included this more with an understanding of protection/confidence...

That is, safe defaults with the option to do riskier things if you indicate you accept that risk.

We're walking a directory tree that is providing data values... shouldn't we ensure that if we're traversing a symlink that the user intends it?

Also, if our intention is to support traversing directories, we've got more work to do here.

pivotaljohn replied:
_(discussed in [Slack 🧵  in #carvel](https://kubernetes.slack.com/archives/CH8KCCKA5/p1653053084082349))_

pivotaljohn replied:
Resolution: keep for now.

jtigger committed:
Make end-to-end tests more navigable

... so that as a contributor, when I read this list of tests, I can
focus my attention to a section of the test file, rather than having to
comprehend the whole thing at once.

- group and sort by purpose
  - feature-specific tests, first
  - input/output mechanism verification, next
  - groups of examples-based tests for verifying the examples, last
- add the "Overlay" Playground example group (currently untested)
- add missing expectations for Overlay examples

gcheadle-vmware reviewed (approved):
I've enjoyed watching this feature evolve into something that uses existing mechanisms in ytt of reading files from a Dir. Nothing stood out to me that was alarming + naming & godocs really helped me understand these changes. 🎊
